File: Data Structures & Algorithms/trapping-rain-water/submission-0.java
class Solution {
    public int trap(int[] height) {
        int total = 0;
        for(int i = 1; i < height.length - 1;i++) {
            int largestLeft = findLargest(height, 0, i - 1);
            int largestRight = findLargest(height, i + 1, height.length - 1);
            if(largestLeft > height[i] && largestRight > height[i]) {
                total += Math.min(largestLeft, largestRight) - height[i];
            }
        }

        return total;
    }

    private int findLargest(int[] height, int l, int r) {
        int max = 0;
        while(l<=r) {
            max = Math.max(max, height[l]);
            l++;
        }

        return max;
    }
}


/**
2 + 2 + 3 + 2
**/

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-0.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        freq = {}
        n = len(s)
        l, r = 0, 0
        mx = 0

        while r < n:
            freq[s[r]] = freq.get(s[r], 0) + 1
            
            highest_freq = max(freq.values())
            while r - l + 1 - highest_freq > k:
                freq[s[l]] = freq[s[l]] - 1
                l+=1
                highest_freq = max(freq.values())
            mx = max(mx, r - l + 1)
            r+=1 
        return mx    

        



